### DISCUSSION TITLE
@quasar/cli - undefined

### DISCUSSION THREAD

**Discussion by anazzani** [quasarframework/quasar#9062] [CLI - BEX mode]
Hi all.

I just [posted](https://forum.quasar-framework.org/topic/7837/quasar-cli-undefined) in the old forums but now I noticed that this is probably the most appropriate place for my request.

So I'll repeat my question here: can somebody help me fix this problem?
```
$ quasar info

Operating System - Darwin(20.3.0) - darwin/x64
NodeJs - 14.16.1

Global packages
  NPM - 6.14.12
  yarn - 1.22.10
  @quasar/cli - undefined
  @quasar/icongenie - 2.3.3
  cordova - Not installed

Important local packages
  quasar - 2.0.0-beta.13 -- Build high-performance VueJS user interfaces (SPA, PWA, SSR, Mobile and Desktop) in record time
  @quasar/app - 3.0.0-beta.17 -- Quasar Framework local CLI
  @quasar/extras - 1.10.4 -- Quasar Framework fonts, icons and animations
  eslint-plugin-quasar - Not installed
  vue - 3.0.11 -- vue
  vue-router - 4.0.5
  vuex - 4.0.0 -- state management for Vue.js
  electron - Not installed
  electron-packager - Not installed
  electron-builder - Not installed
  @babel/core - 7.13.16 -- Babel compiler core.
  webpack - 5.35.0 -- Packs CommonJs/AMD modules for the browser. Allows to split your codebase into multiple bundles, which can be loaded on demand. Support loaders to preprocess files, i.e. json, jsx, es7, css, less, ... and your custom stuff.
  webpack-dev-server - 4.0.0-beta.2 -- Serves a webpack app. Updates the browser on changes.
  workbox-webpack-plugin - Not installed
  register-service-worker - 1.7.2 -- Script for registering service worker, with hooks
  typescript - 4.2.2 -- TypeScript is a language for application scale JavaScript development
  @capacitor/core - Not installed
  @capacitor/cli - Not installed
  @capacitor/android - Not installed
  @capacitor/ios - Not installed

Quasar App Extensions
  *None installed*

$ quasar upgrade
 App • Looking for Quasar App Extension "upgrade" command
 App • ⚠️  Quasar App Extension "upgrade" is missing...
```

I tried the above command both in the default macOS terminal and in iTerm2.

Thanks.

**Comment by ontwikkelfabriek** [comment]
>  @quasar/cli - undefined

This is your problem, Quasar CLI  is not installed correctly globally. But I guess you already figured that out. 

Looking at your install commands from the old forum, could it be as simple as misplacing the npm **-g** option, `npm install -g @quasar/cli` instead of `npm -g install  @quasar/cli`?

See:
https://docs.npmjs.com/downloading-and-installing-packages-globally

**Reply by anazzani** [reply]
I followed the instructions [here](https://next.quasar.dev/quasar-cli/installation) and used the command `yarn global add @quasar/cli`

In fact currently I see:
```
$ yarn global list |grep cli
info "@quasar/cli@1.1.3" has binaries:
```

Is that correct?

**Reply by ontwikkelfabriek** [reply]
globals libs are (used to be)  highly recommended to be installed by **npm** instead of yarn. Anyways the `@quasar/cli - undefined` is your problem.... Seeing you have this problem for a long time now ,maybe you could try another machine or VM? Besides that I am all out of ideas how to fix your problem.....

**Reply by anazzani** [reply]
Thanks for your help.

Unfortunately using npm seems to make no difference. After
```
$ yarn global remove @quasar/cli
yarn global v1.22.10
[1/2] 🗑  Removing module @quasar/cli...
[2/2] 🔨  Regenerating lockfile and installing missing dependencies...
success Uninstalled packages.
✨  Done in 1.21s.

$ npm i -g @quasar/cli
npm WARN deprecated coffee-script@1.12.7: CoffeeScript on NPM has moved to "coffeescript" (no hyphen)
/Users/archimede/.nvm/versions/node/v14.16.1/bin/quasar -> /Users/archimede/.nvm/versions/node/v14.16.1/lib/node_modules/@quasar/cli/bin/quasar
+ @quasar/cli@1.1.3
added 377 packages from 292 contributors in 30.055s
```
the cli is still undefined.

Could my problem be [nvm](https://github.com/nvm-sh/nvm)?

And maybe a stupid question, but do I really _need_ the quasar/cli? Is there an alternative way to upgrade a quasar project?

For example manually modifying the package.json of my app (updating the version of @quasar/extras - @quasar/app - quasar when needed) will have the exact SAME effect of the `quasar upgrade` command?

**Reply by ontwikkelfabriek** [reply]
> Could my problem be nvm?

One way to find out: remove it and do a plain npm install. I use nvm on windows and I don't have this issue...

> Is there an alternative way to upgrade a quasar project? For example manually modifying the package.json of my app (updating the version of @quasar/extras - @quasar/app - quasar when needed) 

I think that might work. I would recommend reading the change logs for the version you are updating **to**. You can always create a new project with the new version and do a compare to your manually updated version.

**Reply by anazzani** [reply]
Thanks for your support.

Out of desperation I tried again the commands posted in the old thread and it worked!

Not sure what changed: I had tried those command before creating this thread to no avail.

Anyway, I'm re-posting them in case somebody else experiences the same issue:
```
npm -g cache clean --force
npm -g cache verify
npm cache clean --force
npm cache verify
npm -g uninstall @quasar/cli
npm -g install @quasar/cli
npm uninstall @quasar/cli
rm -fr node_modules
rm package-lock.json
npm i
```
Cheers.

**Reply by hawkeye64** [reply]
Yes, you don't really need the global cli except for creating a new project or serving.

You can add commands to the scripts section of your package.json
`"upgrade": "quasar upgrade -i"`
Then `yarn upgrade` - and yarn will find your local quasar (app)

**Reply by hawkeye64** [reply]
Alternatively, you can use `npx`
